intellij/src/saros/intellij/filesystem/IntelliJProjectImpl.java

```diff
@@ -523,9 +527,28 @@ public VirtualFile findVirtualFile(final IPath path) {
     return null;
   }
 
+  /**
+   * {@inheritDoc}
+   *
+   * <p>This method operates under the assumption that disposed {@link Project} objects still allow
+   * requests for the project name. This method needs to be adjusted once this assumption no longer
+   * holds.
+   *
+   * @return a hash code value for this object
+   */
   @Override
   public int hashCode() {
-    return moduleName.hashCode();
+    String projectName = project.getName();
+    String projectFilePath;
+
+    if (!project.isDisposed()) {
```

Comment:
Hmmm. I just found out that this causes issues when the class is used in combination with hash-comparison-based data structures like hash sets. E.g. using `HashSet<T>.contains(T)` no longer works consistently as it compares the initial hash to the current hash, which can be different if the contained project object is disposed (or the project file path changed).

I am not sure what to do about this. Requiring ever other part of the code using other set implementations seems ...bad... as the hash-based implementations are usually the ones used by default and this issue is not directly apparent when using the class.

Reply:
This is also an issue with the current eclipse implementation. The resource hash code of the Eclipse sources is solely base on the path of the resource, meaning if the path changes, the hashcode changes. This leads to issues when moving partially shared resources because of [this](https://github.com/saros-project/saros/blob/master/core/src/saros/session/internal/SharedProjectMapper.java#L307) check.

This is an even bigger issue if we are allowing shared projects to be moved during the session as the completely shared projects are also held as a HashSet.

@srossbach @m273d15 @stefaus Any idea how to handle this nicely, preferably without going through the whole codebase and checking the usage of hash sets (especially when used for long-term data storage)?

Reply:
I do not get the Eclipse stuff, see https://github.com/saros-project/saros/blob/master/core/src/saros/session/internal/SarosSession.java#L887

As for the hashcode. It does not matter what hashcode you return, however it has to be always the same value.

h(a) = h(b) does not imply that a = b, however if a = b => h(a) = h(b)

You simply could return 0 although this would mess up the hashmap. So I guess returning the hashed project name is fine.

Reply:
And BTW all Eclipse resource implementations are immutable so you cannot alter the hashcode in any way.

Reply:
> I do not get the Eclipse stuff, see https://github.com/saros-project/saros/blob/master/core/src/saros/session/internal/SarosSession.java#L887

Ah, ok. I was not aware of that part of the logic.

> As for the hashcode. It does not matter what hashcode you return, however it has to be always the same value.

I interpreted the javadoc of `Object.hashCode` to mean that it always has to return the same value *as long as the state used for equals does not change*. But you are right, just using the name is more sensible as hash collisions in case where there are multiple project module combinations with the same name don't matter.

> And BTW all Eclipse resource implementations are immutable so you cannot alter the hashcode in any way.

Hmm, I guess I was a bit quick to conclude that the issue (moving stuff in a partial session does not work) is with the hash values.